tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/passes/TemporalPassImpl.scala

```diff
@@ -0,0 +1,133 @@
+ * formula. The encoding is described in https://lmcs.episciences.org/2236, Sections 3.2 and 4.
+ */
+class TemporalPassImpl @Inject() (
+    val options: PassOptions,
+    tracker: TransformationTracker,
+    gen: UniqueNameGenerator,
+    writerFactory: TlaWriterFactory)
+    extends TemporalPass with LazyLogging {
+
+  override def name: String = "TemporalPass"
+
+  override def execute(tlaModule: TlaModule): PassResult = {
+    logger.info("  > Rewriting temporal operators...")
+
+    val newModule = options.get[List[String]]("checker", "inv") match {
+      case None =>
+        logger.info("  > No invariants specified, nothing to encode")
+        tlaModule
+      case Some(invariants) =>
+        val init = options.get[String]("checker", "init")
+        if (init.isEmpty) {
+          logger.info("  > `init` is not set, cannot encode invariants")
+          None
+        }
+        val next = options.get[String]("checker", "next")
+        if (init.isEmpty) {
+          logger.info("  > `next` is not set, cannot encode invariants")
+          None
+        }
+        temporalToInvariants(tlaModule, invariants, init.get, next.get)
+    }
+
+    writeOut(writerFactory, newModule)
+
+    Right(newModule)
+  }
+
+  
+  def temporalToInvariants(
+      module: TlaModule,
+      temporalProperties: List[String],
+      init: String,
+      next: String): TlaModule = {
+    val levelFinder = new TlaLevelFinder(module)
+
+    val temporalFormulas = temporalProperties
+      .map(invName => {
+        module.declarations.find(_.name == invName)
+      })
+      .filter(invOption =>
+        invOption match {
+          case Some(inv: TlaOperDecl) if inv.formalParams.isEmpty =>
+            // either a state invariant, or an action invariant
+            val level = levelFinder.getLevelOfDecl(inv)
+            level == TlaLevelTemporal
+          case _ => false
+        })
+      .map(invOption => invOption.get.asInstanceOf[TlaOperDecl])
+
+    if (temporalFormulas.isEmpty) {
+      logger.info("  > No temporal properties found, nothing to encode")
+      module
+    } else {
+      logger.info(s"  > Found ${temporalFormulas.length} temporal temporal properties")
+      logger.info(s"  > Adding logic for loop finding")
+
+      val initDecl = module.declarations.find(_.name == init) match {
+        case Some(initDecl: TlaOperDecl) =>
+          val nparams = initDecl.formalParams.length
+          if (nparams != 0) {
+            val message =
+              s"Expected init predicate $init to have 0 params, found $nparams parameters"
+            throw new TlaInputError(message, Some(initDecl.body.ID))
+          }
+          initDecl
+        case None =>
+          val message = (s"init predicate named `${init}` not found")
+          throw new TlaInputError(message)
+        case _ =>
+          val message = (s"Expected to find a predicate named `${init}` but did not")
+          throw new TlaInputError(message)
+      }
```

Comment:
No, not really. This was more or less only as I didn't quite know whether this would all be guaranteed to be fine, but I never ran into any issues while testing, so I've removed all checking that init and next are fine, assuming another pass before this handles it if not